File: LeetCode/0622-design-circular-queue/0622-design-circular-queue.java
class MyCircularQueue {
    
    int[] cq;
    int index;
    int front;
    int rear;

    public MyCircularQueue(int k) {
        cq = new int[k];
        index = 0;
        front = 0;
        rear = 0;

        Arrays.fill(cq, -1);
    }
    
    public boolean enQueue(int value) {
        if (isFull()) return false;
        
        cq[index] = value;
        rear = index;
        
        index++;

        if (index == cq.length) index = 0;
        
        return true;
    }
    
    public boolean deQueue() {
        if (isEmpty()) return false;
        
        cq[front] = -1;
        
        if (front == cq.length - 1) front = 0;
        else front++;
        
        return true;
    }
    
    public int Front() {
        return isEmpty() ? -1 : cq[front];
    }
    
    public int Rear() {
        return isEmpty() ? -1 : cq[rear];
    }
    
    public boolean isEmpty() {
        return cq[front] == -1;
    }
    
    public boolean isFull() {
        if (isEmpty()) return false;

        return front == 0 ? rear == cq.length - 1 : rear == front - 1;
    }
}

File: LeetCode/0622-design-circular-queue/0622-design-circular-queue.kt
class MyCircularQueue(k: Int) {
    var q = IntArray(k)
    var front = 0
    var rear = -1
    var len = 0

    fun enQueue(value: Int): Boolean {
        if (isFull()) return false 
        
        rear = ++rear % q.size
        q[rear] = value
            
        len++
        
        return true
    }

    fun deQueue(): Boolean {
        if (isEmpty()) return false
        
        front = ++front % q.size
            
        len--
            
        return true
    }

    fun Front(): Int {
        if (isEmpty()) return -1
        
        return q[front]
    }

    fun Rear(): Int {
        if (isEmpty()) return -1
        
        return q[rear]
    }

    fun isEmpty(): Boolean {
        return len == 0
    }

    fun isFull(): Boolean {
        return len == q.size
    }

}